README.md
# About interactionR

[![R-CMD-check](https://github.com/tunsmart/interactionR/actions/workflows/check-standard.yaml/badge.svg)](https://github.com/tunsmart/interactionR/actions/workflows/check-standard.yaml)
[![](https://www.r-pkg.org/badges/version/interactionR)](https://cran.r-project.org/package=interactionR)
![Downloads](https://cranlogs.r-pkg.org/badges/interactionR)
[![](http://cranlogs.r-pkg.org/badges/grand-total/interactionR)](https://cran.r-project.org/package=interactionR)


InteractionR allows researchers to produce publication-ready tables that
includes all effect estimates necessary for full reporting effect
modification and interaction analysis as recommended by Knol and
Vanderweele (2012) (doi:10.1093/ije/dyr218). It also estimates
confidence interval for the trio of additive interaction measures using
the delta method (Hosmer and Lemeshow (1992),
doi:10.1097/00001648-199209000-00012), variance recovery method
(Zou (2008), doi:10.1093/aje/kwn104), or percentile bootstrapping
(Assmann et al. (1996), doi:10.1097/00001648-199605000-00012).

# Installation

interactionR is on
[CRAN](https://cran.r-project.org/package=interactionR), install using
`install.packages("interactionR")`.

You can also install the development version from
[GitHub](https://github.com/tunsmart/interactionR) with:

``` r
# install.packages("devtools")
devtools::install_github("tunsmart/interactionR")
```

# Implementation

interactionR is implemented in the R statistical software environment; R
is installable from <https://www.r-project.org/>. The main user-facing
function of the package, `interactionR()`, accepts the following
arguments:

  - *model*: a regression model fitted by the user with interaction term
    for the two exposures under consideration. This may be an object of
    class glm with a valid link for logistic regression or approximants
    of risk ratio, class clogit or class coxph. It can also include
    confounders adjustment as is usually the case.

  - *exposure\_names*: A character vector of two named binary exposure
    variables present in the fitted model.

  - *ci.type*: A character string (“delta” or “mover”) specifying the
    method to use for the estimation of CI for the measures of additive
    interaction. Default is “delta”.

  - *ci.level*: Magnitude of the returned CI level.

  - *em*: TRUE, for effect modification assessment. FALSE, for
    interaction.

  - *recode*: If TRUE, recodes the exposures - if at least one of the
exposures is protective - such that the stratum with the lowest risk
becomes the new reference category when the two exposures are considered
jointly (See Knol et al (2011) \[doi: 10.1007/s10654-011-9554-9\]).

# Example: The joint effect of alcohol and smoking on oral cancer.

Consider the case-control data from Rothman and Keller (1979) \[doi:
10.1016/0021-9681(72)90006-9\], which studied the joint effect of
alcohol and smoking on oral cancers. This dataset was used previously by
Hosmer and Lemeshow, and Zou and its included in our package as ‘OCdata’
for easy replication of our analysis by the user. It contains the two
exposures, alcohol (‘alc’) and smoking (‘smk’), and the outcome, oral
cancer (‘oc’) as dichotomous variables (0 or 1, indicating absence or
presence respectively). Suppose we are considering interaction and we
want to compute the CIs for the measures of additive interaction using
the MOVER method, we will start by fitting the following logistic
regression model with an interaction term for alcohol and smoking on
oral cancer:

``` r
library(interactionR)
data (OCdata)

## fit the interaction model
model.glm <- glm(oc ~ alc*smk, family = binomial(link = "logit"), data = OCdata)
```

Then, pass the fitted model to the function which generates a list
object of class ‘interactionR’

``` r
table_object = interactionR(model.glm, exposure_names = c("alc", "smk"), ci.type = "mover", ci.level = 0.95, em = F, recode = F)
```

This returns a list object of class interactionR which includes a
dataframe containing all the effect estimates that are necessary to
fully report effect modification or interaction. This data frame and
other components of the list are accessible to the user for further
manipulation, if desired. Importantly, the output object is formatted in
such a way that the data frame can be processed by the tabling function
`interactionR_table()`. An example is:

``` r
interactionR_table(table_object)
```

<img src="man/figures/Figure-1.png" width="100%" />

The tabling function will generate and save a publication-ready table as
a word document to the user’s working directory, if desired. The
returned point estimate and CI for RERI (3.74; 95% CI: -11.43, 21.87) is
as reported by Zou for this data (13) – If the ‘ci.type’ argument in the
`interactionR()` call had been set to “delta”, the CIs for this trio of
interaction measures would be as reported by Hosmer and Lemeshow.

In addition to the main function described above, the package also
provides standalone functions for each of the CI estimating methods:
interactionR\_mover(), interactionR\_delta(), and interactionR\_boot().
The latter implements the percentile bootstrapping of CIs of additive
interaction measures as described by Assmann et al, with usage shown in
the next example.

# Example 2: Effect of sports participation and smoking on herniated lumbar disc.

To illustrate the interactionR\_boot() function, consider the
case-control data of the effect of sports participation and smoking on
herniated lumbar disc examined by Assmann et al. in their analysis. The
dataset is available in the package as ‘HDiscdata’ and contains three
binary variables: i) the outcome. herniated lumbar disc, ‘h’; and ii &
iii) the exposures sports participation, ‘ns’, and smoking ‘smk’. The
function accepts the following arguments:

  - *model*: A fitted model of class glm. Requires the exposures with
    interaction term to be listed first before any other
    covariates/confounders (if applicable).

  - *ci.level*; *em*; *recode*: As previously described for
    `interactionR()`.

  - *seed*: The random number seed to use for generating the bootstrap
    samples (for reproducibility). Default is 12345 but can be set to
    any number.

  - *s*: Number of bootstrap resampling. Default is 1000

Again, we start by fitting a logistic regression model:

``` r
data(HDiscdata)
m2 = glm(h ~ ns*smk, family = binomial(link = 'logit'), data = HDiscdata)
```

Then, pass the object to the `interactionR_boot()` function as follows:

``` r
table_object2 = interactionR_boot(m2, ci.level = 0.95, em = F, recode = F, seed = 12345, s = 1000)
```

    Loading required namespace: boot

This runs a non-parametric bootstrap sample 1000 times with replacements
and a percentile CI. The function also returns a list object of class
interactionR containing all the desired estimates and manipulable by the
tabling function - `interactionR_table()` - described earlier. Calling
`interactionR_table()` on the returned object produces a
publication-ready table with estimates for the CI of RERI and AP similar
to that reported by Assmann et al. for this data.

``` r
interactionR_table(table_object2)
```

<img src="man/figures/Figure-2.png" width="100%" />
Furthermore, some base R functions are available to the user to further
manipulate some parts of the output object from interactionR\_boot(). A
simple example is:

``` r
hist(table_object2$bootstrap)
```

<img src="man/figures/README-example7-1.png" width="100%" />

This produce histograms of the distribution of each of the three
bootstrapped parameters (RERI, AP and SI), allowing the user to inspect
the overall performance and accuracy of the returned estimates.

# To cite
Cite as: Babatunde Y. Alli (2021). InteractionR: An R package for full reporting of effect modification and interaction. Software Impacts, 10, 100147. doi:10.1016/j.simpa.2021.100147

# Community guidelines

### To contribute

For contribute to interactionR please feel free to fork the branch and
send me pull requests. Currently, only the master branch is used for
both development and releases.

### To report problems, bugs or issues

Please use [interactionR
issues](https://github.com/tunsmart/interactionR/issues) here on Github.

### To get support

Please consult the help files for the package within R or contact
the maintainer at babatunde.alli \[at\] mail.mcgill.ca




NEWS.md
interactionR 0.1.7
==================

*Patch to correct pvalue estimation for SI and relax argument checkings

interactionR 0.1.6
==================

*Bug fix to correct pvalue estimation for survival objects 

interactionR 0.1.5
==================

*Quick patch for the last update. Made a correction on an important information for RERI in the documentation.

interactionR 0.1.4
==================

*Numerous bug fixes
*Made recoding optional when at least one exposure is preventive

interactionR 0.1.3
==================

*adds optional pvalue outputting feature to the tabling function

interactionR 0.1.2
==================
Patched version

*Bug fix for wrong estimation of the CI for the joint effects in 'coxph' and 'clogit' models

interactionR 0.1.1
==================
Patched version

*The dependency Huxtable is now replaced with Flextable


interactionR 0.1.0
==================

First release


R/emm.R
#' @export
#' @importFrom msm deltamethod
emm = function(model, exposure_names = c(), ci.type = "delta", ci.level = 0.95, 
    em = T, recode = F) {
    
    if (invalid(model)) {
        stop("The 'model' argument must be a regression model object fit with glm() and link = 'logit', coxph() or clogit()")
    } else if (class(exposure_names) != "character") {
        stop("Argument 'exposure_names' requires a character vector of the names of the two exposure variables ")
    }
    
    # Estimates the critical value from the supplied CI.level for
    # subsequent CI estimations
    alpha = 1 - ci.level
    z = qnorm(1 - alpha/2)
    
    # Extracts the names for the main exposure (beta1), the effect modifier
    # (beta2) and their joint efffect (beta1 + beta2 + beta1:beta2)
    beta1 = exposure_names[1]
    beta2 = exposure_names[2]
    beta3 = paste(beta1, beta2, sep = ":")
    
    varNames = c(beta1, beta2, beta3)
    
    # estimating coefficients to check for any preventive exposures
    b1 = coef(model)[beta1]
    b2 = coef(model)[beta2]
    b3 = coef(model)[beta3]
    
    
    # check if any exposure is preventive
    if (preventive(OR10 = exp(b1), OR01 = exp(b2))) {
        if (!recode) 
            stop("Error: At least one exposure is preventive. Set argument recode=TRUE for the exposures to be automatically recoded. see Knol et al. (2011) European Journal of Epidemiology, 26(6), 433-438")
        if (recode) {
            # find stratum with lowest overall risk, conditional on covariates
            temp = data.frame(cat = c("OR10", "OR01", "OR11"), value = c(exp(b1), 
                exp(b2), exp(b1 + b2 + b3)))
            refcat = temp$cat[which.min(temp$value)]  # get category's name
            
            # extract first 'subscript' number
            E1.ref = substr(refcat, 3, 3)
            # extract second 'subscript' number
            E2.ref = substr(refcat, 4, 4)
            
            # recode each exposure based on new reference category
            dat = model$data
            dat[[beta1]] = ifelse(dat[[beta1]] == E1.ref, 0, 1)
            dat[[beta2]] = ifelse(dat[[beta2]] == E2.ref, 0, 1)
            
            # inform the user
            warning("Recoding exposures; new reference category for ", 
                beta1, " is ", E1.ref, " and for ", beta2, " is ", E2.ref)
            
            # refit model with user's original call but recoded data
            model = update(model, . ~ ., data = dat)
            
            # get new coefficients and ORs
            b1 = coef(model)[beta1]
            b2 = coef(model)[beta2]
            b3 = coef(model)[beta3]
            
        }
    }
    
    se_vec = summary(model)$coefficients[, 2]  #extracts the SE vector for the coefficients
    # from the model
    v1 = se_vec[beta1]^2
    v2 = se_vec[beta2]^2
    v3 = se_vec[beta3]^2
    
    ### Extracts the variance-covariance matrix from the model### for use in
    ### the delta method CI estimation for RERI and AP###
    v_cov = vcov(model)
    v_cov1 = v_cov[varNames, varNames]  #for deltamethod
    cov12 = v_cov[beta1, beta2]
    cov13 = v_cov[beta1, beta3]
    cov23 = v_cov[beta2, beta3]
    v123 = v1 + v2 + v3 + (2 * (cov12 + cov13 + cov23))
    v12 = v1 + v2 + (2 * (cov12))
    v13 = v1 + v3 + (2 * (cov13))
    v23 = v2 + v3 + (2 * (cov23))
    
    
    # Estimates individual and joint effects ORs (with CI) from the model
    
    OR00 = 1  #reference OR
    OR10 = as.numeric(exp(b1))
    l1 = exp(confint.default(model)[beta1, 1])
    u1 = exp(confint.default(model)[beta1, 2])  # This is also OR of X on D (A==0)
    OR01 = as.numeric(exp(b2))
    l2 = exp(confint.default(model)[beta2, 1])
    u2 = exp(confint.default(model)[beta2, 2])  # This is also OR of A on D (X==0)
    OR11 = as.numeric(exp(b1 + b2 + b3))
    l3 = exp(b1 + b2 + b3 - z * sqrt(v123))
    u3 = exp(b1 + b2 + b3 + z * sqrt(v123))
    
    ### Estimates the effect (and CI) of A on D (X==1) ###
    OR_X1 = as.numeric(exp(b2 + b3))  # OR of A on D (X==1)
    CI.ll_OR_X1 = exp(b2 + b3 - z * sqrt(v23))
    CI.ul_OR_X1 = exp(b2 + b3 + z * sqrt(v23))
    
    
    ### Estimates the effect (and CI) of X on D (A==1) ###
    OR_A1 = as.numeric(exp(b1 + b3))  # OR of X on D (A==1)
    CI.ll_OR_A1 = exp(b1 + b3 - z * sqrt(v13))
    CI.ul_OR_A1 = exp(b1 + b3 + z * sqrt(v13))
    
    # Effect modification on the multiplicative scale and CI
    OR_M = as.numeric(exp(b3))
    CI.ll_OR_M = exp(confint.default(model)[beta3, 1])
    CI.ul_OR_M = exp(confint.default(model)[beta3, 2])
    
    if (ci.type == "mover") {
        # Estimates measures of effect modification on the additive scale and
        # calculates their CI with the 'MOVER' method (Zou (2018)
        # https://doi.org/10.1093/aje/kwn104)
        
        # RERI, CI and p-value
        RERI = OR11 - OR01 - OR10 + 1
        r12 = (v1 + cov12 + cov13)/sqrt(v1 * v123)
        r13 = (cov12 + v2 + cov23)/sqrt(v2 * v123)
        r23 = cov12/sqrt(v1 * v2)
        
        p1 = (OR11 - l3)^2 + (u1 - OR10)^2 + (u2 - OR01)^2
        p2 = 2 * r12 * (OR11 - l3) * (u1 - OR10)
        p3 = 2 * r13 * (OR11 - l3) * (u2 - OR01)
        p4 = 2 * r23 * (u1 - OR10) * (u2 - OR01)
        p5 = p1 - p2 - p3 + p4
        p6 = p5^0.5
        
        L = 1 + OR11 - OR10 - OR01 - p6
        
        k1 = (u3 - OR11)^2 + (OR10 - l1)^2 + (OR01 - l2)^2
        k2 = 2 * r12 * (u3 - OR11) * (OR10 - l1)
        k3 = 2 * r13 * (u3 - OR11) * (OR01 - l2)
        k4 = 2 * r23 * (OR10 - l1) * (OR01 - l2)
        k5 = (k1 - k2 - k3 + k4)^0.5
        
        U = 1 + OR11 - OR10 - OR01 + k5
        
        
        # AP, CI and p-value
        theta1 = 1/exp(b1 + b2 + b3)
        theta2 = 1/exp(b2 + b3)
        theta3 = 1/exp(b1 + b3)
        AP = theta1 - theta2 - theta3 + 1
        APr12 = (cov12 + cov13 + v2 + (2 * cov23) + v3)/sqrt(v23 * v123)
        APr13 = (v1 + cov12 + (2 * cov13) + cov23 + v3)/sqrt(v13 * v123)
        APr23 = (cov12 + cov23 + cov13 + v3)/sqrt(v23 * v13)
        
        APl1 = theta1 * exp(-z * sqrt(v123))
        APu1 = theta1 * exp(z * sqrt(v123))
        
        APl2 = theta2 * exp(-z * sqrt(v23))
        APu2 = theta2 * exp(z * sqrt(v23))
        
        APl3 = theta3 * exp(-z * sqrt(v13))
        APu3 = theta3 * exp(z * sqrt(v13))
        
        APp1 = (theta1 - APl1)^2 + (APu2 - theta2)^2 + (APu3 - theta3)^2
        APp2 = 2 * APr12 * (theta1 - APl1) * (APu2 - theta2)
        APp3 = 2 * APr13 * (theta1 - APl1) * (APu3 - theta3)
        APp4 = 2 * APr23 * (APu2 - theta2) * (APu3 - theta3)
        APp5 = APp1 - APp2 - APp3 + APp4
        APp6 = APp5^0.5
        
        APL = 1 + theta1 - theta2 - theta3 - APp6
        
        APk1 = (APu1 - theta1)^2 + (theta2 - APl2)^2 + (theta3 - APl3)^2
        APk2 = 2 * APr12 * (APu1 - theta1) * (theta2 - APl2)
        APk3 = 2 * APr13 * (APu1 - theta1) * (theta3 - APl3)
        APk4 = 2 * APr23 * (theta2 - APl2) * (theta3 - APl3)
        APk5 = (APk1 - APk2 - APk3 + APk4)^0.5
        
        APU = 1 + theta1 - theta2 - theta3 + APk5
        
        
        # SI, CI and p-value
        SItheta1 = log((exp(b1 + b2 + b3) - 1))
        SItheta2 = log((exp(b1) + exp(b2) - 2))
        lnSI = SItheta1 - SItheta2
        SI = exp(lnSI)
        
        vSItheta1 = (exp(b1 + b2 + b3)/(exp(b1 + b2 + b3) - 1))^2 * v123
        vSItheta2 = ((exp(2 * b1) * v1) + (exp(2 * b2) * v2) + (2 * exp(b1 + 
            b2) * cov12))/(exp(b1) + exp(b2) - 2)^2
        SIl1 = SItheta1 - z * sqrt(vSItheta1)
        SIu1 = SItheta1 + z * sqrt(vSItheta1)
        SIl2 = SItheta2 - z * sqrt(vSItheta2)
        SIu2 = SItheta2 + z * sqrt(vSItheta2)
        
        SIr = ((exp(b1) * (v1 + cov12 + cov13)) + (exp(b2) * (cov12 + v2 + 
            cov23)))/sqrt(v123 * ((exp(2 * b1) * v1) + (exp(2 * b2) * v2) + 
            (2 * exp(b1 + b2) * cov12)))
        
        lnSIL = (SItheta1 + (-SItheta2)) - sqrt((SItheta1 - SIl1)^2 + ((-SItheta2) - 
            (-SIl2))^2 + (2 * SIr * (SItheta1 - SIl1) * ((-SItheta2) - 
            (-SIl2))))
        lnSIU = (SItheta1 + (-SItheta2)) + sqrt((SIu1 - SItheta1)^2 + ((-SIu2) - 
            (-SItheta2))^2 + (2 * SIr * (SIu1 - SItheta1) * ((-SIu2) - 
            (-SItheta2))))
        SIL = exp(lnSIL)
        SIU = exp(lnSIU)
    } else if (ci.type == "delta") {
        # Estimates measures of effect modification on the additive scale and
        # calculates their CI and p-value with the delta method implemented in
        # the msm package
        
        # RERI, CI and p-value
        RERI = OR11 - OR01 - OR10 + 1
        se_RERI = deltamethod(g = ~exp(x1 + x2 + x3) - exp(x1) - exp(x2) + 
            1, mean = c(b1, b2, b3), cov = v_cov1)
        L = RERI - z * se_RERI
        U = RERI + z * se_RERI
        
        
        # AP, CI and p-value
        AP = RERI/OR11
        se_AP = deltamethod(g = ~(exp(x1 + x2 + x3) - exp(x1) - exp(x2) + 
            1)/exp(x1 + x2 + x3), mean = c(b1, b2, b3), cov = v_cov1)
        APL = AP - z * se_AP
        APU = AP + z * se_AP
        
        
        
        # SI, CI and p-value
        lnSI = log((exp(b1 + b2 + b3) - 1)) - log((exp(b1) + exp(b2) - 
            2))
        SI = exp(lnSI)
        se_SI = deltamethod(g = ~log((exp(x1 + x2 + x3) - 1)) - log((exp(x1) + 
            exp(x2) - 2)), mean = c(b1, b2, b3), cov = v_cov1)
        
        SIL = exp(lnSI - z * se_SI)
        SIU = exp(lnSI + z * se_SI)
        
    } else {
        stop("Argument 'ci.type' must be 'delta' or 'mover' ")
    }
    
    d = data.frame(Measures = c("OR00", "OR01", "OR10", "OR11", paste("OR(", 
        beta2, " on outcome [", beta1, "==0]", sep = ""), paste("OR(", 
        beta2, " on outcome [", beta1, "==1]", sep = ""), "Multiplicative scale", 
        "RERI"), Estimates = c(OR00, OR01, OR10, OR11, OR01, OR_X1, OR_M, 
        RERI), CI.ll = c(NA, l2, l1, l3, l2, CI.ll_OR_X1, CI.ll_OR_M, L), 
        CI.ul = c(NA, u2, u1, u3, u2, CI.ul_OR_X1, CI.ul_OR_M, U))
    rownames(d) = NULL
    
    
    
    if (!em) {
        d = data.frame(Measures = c("OR00", "OR01", "OR10", "OR11", paste("OR(", 
            beta2, " on outcome [", beta1, "==0]", sep = ""), paste("OR(", 
            beta2, " on outcome [", beta1, "==1]", sep = ""), paste("OR(", 
            beta1, " on outcome [", beta2, "==0]", sep = ""), paste("OR(", 
            beta1, " on outcome [", beta2, "==1]", sep = ""), "Multiplicative scale", 
            "RERI", "AP", "SI"), Estimates = c(OR00, OR01, OR10, OR11, 
            OR01, OR_X1, OR10, OR_A1, OR_M, RERI, AP, SI), CI.ll = c(NA, 
            l2, l1, l3, l2, CI.ll_OR_X1, l1, CI.ll_OR_A1, CI.ll_OR_M, L, 
            APL, SIL), CI.ul = c(NA, u2, u1, u3, u2, CI.ul_OR_X1, u1, CI.ul_OR_A1, 
            CI.ul_OR_M, U, APU, SIU))
        rownames(d) = NULL
    }
    
    tabler(d = d, beta1 = beta1, beta2 = beta2, em = em)
    
}
